twine/package.py

```diff
@@ -51,12 +51,12 @@
 def _safe_name(name: str) -> str:
     """Convert an arbitrary string to a standard distribution name.
 
-    Any runs of non-alphanumeric/. characters are replaced with a single '-'.
+    Any runs of non-alphanumeric/./_ characters are replaced with a single '-'.
 
     Copied from pkg_resources.safe_name for compatibility with warehouse.
```

Comment:
More that this comes from elsewhere and is intended to match behavior with PyPI (the server we care most about). This ostensibly breaks that compatibility

Reply:
I disagree. Pls check https://packaging.python.org/en/latest/guides/distributing-packages-using-setuptools/#name 

> Comparison of project names is case insensitive and treats arbitrarily-long runs of underscores, hyphens, and/or periods as equal. For example, if you register a project named cool-stuff, users will be able to download it or declare a dependency on it using any of the following spellings:

> Cool-Stuff
> cool.stuff
> COOL_STUFF
> CoOl__-.-__sTuFF

So underscores and hyphens shouldn't matter in PyPI and it should be backward compatible.
As well this PR is just correct implementation of [PEP 508](https://peps.python.org/pep-0508/).

Reply:
I think there has been some recent work on this, and possibly some changes. I want to dig into that before reviewing. That said, the linked guide is outdated; https://setuptools.pypa.io/en/stable/ is the authoritative documentation for setuptools.

Reply:
After some initial research, I don't know what's the right thing to do, and it seems like there's ongoing discussion about package/distribution name normalization:

- A sub-thread in [Amending PEP 427 (and PEP 625) on package normalization rules](https://discuss.python.org/t/amending-pep-427-and-pep-625-on-package-normalization-rules/17226/52) on the Packaging forum
- The issue that led to the addition of `_safe_name`: [Namespace packages with dots in them fail to upload to PyPI](https://github.com/pypa/twine/issues/743)

(cc @pfmoore @dstufft @ewdurbin as contributors to the above discussions)

> As well this PR is just correct implementation of [PEP 508](https://peps.python.org/pep-0508/).

It seems like `_safe_name` is attempting to implement [normalization defined in PEP 503](https://peps.python.org/pep-0503/#normalized-names).

For additional context, the issue that led to this PR states:

> This behaviour is causing problems e.g. if you have defined in Artifactory mandatory distribution name prefix containing underscores. E.g. `org_name_prefix`. So if you have package named `org_name_prefix_my_package` it is converted by twine to distribution name `org-name-prefix-my-package` and then Artifactory refuses to publish such package.

I think the important question here is: if Twine *doesn't* convert underscores to hyphens, will that be compatible with PyPI?

Additionally, it makes me a little nervous that Twine has its own normalization function. Has this logic been centralized anywhere?

Reply:
> I think the important question here is: if Twine doesn't convert underscores to hyphens, will that be compatible with PyPI?

I can't answer that, I'm afraid. But I do think that PyPI and twine should agree on either allowing unmodified distribution names (as defined in the project metadata) or using the same normalisation (and IMO PEP 503 is the main contender for a "universal" normalisation rule these days). On the other hand, PyPI should have access to the un-normalised name, so it can respect the project author's wishes in the UI and similar places. I don't have context for this issue, but if the field we're discussing here is the *only* place that the project name is passed to PyPI, I'd say that means that it should be passed unmodified.

My impression is that PyPI has much more difficult compatibility considerations, though, so for practical reasons it might be necessary to "follow what PyPI allows" for the short term. I'd view that as a temporary measure, though.

> Has this logic been centralized anywhere?

[Packaging](https://pypi.org/project/packaging/) has `packaging.utils.canonicalize_name` which (as far as I know) implements PEP 503 normalisation. But the PEP 503 rule is designed to be easy to copy and paste, so I'd consider replicating it to avoid a dependency on `packaging` as perfectly acceptable.

Reply:
I believe neither the Wheel nor the Recent sdist PEP requires the file names to be normalized with PEP 503 normalization. If my memory is correct on that, then I would consider it a bug for PyPI not to support submitting files with those names.

Reply:
I should add, all the above rules are applied with varying degrees of accuracy/consistency across the ecosystem. That needs addressing at some point, but for now it doesn't mean we shouldn't follow them, just that looking at what other tools do is likely to be misleading...

Reply:
Is that one of the changes to the wheel spec that happened without a PEP? I'm just waking up but I seem to remember that, which probably contributed to it.

Reply:
I ask, because PEP 427 never required normalization, just escaping.

Reply:
Yes, if I recall, it was agreed as a "clarification". I don't have a link to the discussion handy, sorry.

Reply:
To clarify: Twine isn't generating filenames. `_safe_name` is used to populate the metadata `name` that's used to upload via the [Warehouse legacy API](https://warehouse.pypa.io/api-reference/legacy.html#upload-api).

https://github.com/pypa/twine/blob/8f5e5d6d42d582ef3ea6ef07da277e0cabd22fd2/twine/package.py#L77

https://github.com/pypa/twine/blob/8f5e5d6d42d582ef3ea6ef07da277e0cabd22fd2/twine/package.py#L140-L149

Where `metadata` is read from an sdist or wheel.

Reply:
In which case I see literally no reason (apart from "PyPI doesn't allow the full range of valid names") for normalising this at all.

Reply:
Oh that.

I wouldn't normalize that at all I don't think. PyPI can handle un-normalized names fine.

Reply:
> I wouldn't normalize that at all I don't think. PyPI can handle un-normalized names fine.

How does that reconcile with https://github.com/pypa/twine/issues/743? From the description, it looks like the metadata name was `mosaik.SimConfig`, which resulted in an error from PyPI, which [still seems to use `pkg_resources.safe_name`](https://github.com/pypi/warehouse/blob/0cb73dadbddbfdab869a79d4e034f59fe6ce9d70/warehouse/forklift/legacy.py#L1133-L1140) to validate the project name and filename.

Reply:
That doesn't appear to be what PyPI was seeing, remember PyPI doesn't read the `METADATA` file, twine does.

The error message there was:

```
Start filename for 'mosaik-simconfig' with 'mosaik-simconfig'.
```

Which comes from the code you linked:

```
"Start filename for {!r} with {!r}.".format(project.name, prefix),
```

The second half of that is the ``safe_name`` prefix, or ``mosaik-simconfig``, but the first half is the ``Project.name``, without any normalization.

That gets set earlier [here](https://github.com/pypi/warehouse/blob/0cb73dadbddbfdab869a79d4e034f59fe6ce9d70/warehouse/forklift/legacy.py#L954-L958).

So it looks like twine was normalizing the project name to ``mosaik-simconfig`` when it was sending the metadata, but was not normalizing the filename ``mosaik.SimConfig-0.1.0rc20210315223057-py2.py3-none-any.whl``, and ``pkg_resources.safe_name`` considers ``.`` a safe character, so it didn't escape it.

I suspect if you just didn't normalize the name metadata, #743 wouldn't have occurred, and indeed #743 identified https://github.com/pypa/twine/commit/0bd26af11b8f97ca97ccb9859693d7a14acf15c3#diff-30ee60b48548622c54dfa266e095021f403ad23ab4848c74b2deed5df329ae00R62 as the commit that introduced the breakage, which is also the commit that changed twine from using ``pkg_resources.safe_name()`` (which escapes names to eliminate bad characters) with ``packaging.utils.canonicalize_name()`` (which normalizes names to their canonical form).

Reply:
Twine doesn't emit html though which is what we print from the server in the original issue description there. I think we might be starting to talk past each other. I don't think twine is at fault for that other issue but was trying to work around it and warn users early?

Reply:
Twine reads the ``METADATA`` file to get the values, chucks them into an HTML form, then ``POSTS`` that to PyPI.

One of the fields of that form is ``name``, which is taken from the ``METADATA`` ``Name`` field.

Previously twine used ``pkg_resources.safe_name()`` to escape that field prior to sending it to PyPI, which turns ``mosaik.SimConfig`` into ``mosaik.SimConfig`` (because none of those characters are unsafe in a filename).

That commit linked changes twine to use ``packaging.utils.canonicalize_name()``, which changes ``mosaik.SimConfig`` into ``mosaik-simconfig``.

So twine is telling PyPI that the name of the project is ``mosaik-simconfig``, when previously it told it that it was ``mosaik.SimConfig``.

Independently, PyPI checks that the filename (which is a different field from the name) starts with the name of the project, which once https://github.com/pypa/twine/pull/736 landed, twine, behind the scenes, renamed ``mosaik.SimConfig`` to ``mosaik-simconfig``, making the filename no longer start with the project name (since it was starting with ``mosaik.SimConfig`` not ``mosaik-simconfig``.

So both #743 and #920 ultimately were caused by #736, which changed twine from simply escaping the project name before sending it to PyPI, to normalizing it.

Twine needs to either:

1. Stop normalizing the project name, allowing projects control over their projects name.
2. Continue to normalize the project name, but also normalize the file name.

Fundamentally, this problem is because twine is sending the same data (the name of the projected) differently in two different contexts (context 1: The "name" metadata, context 2: the start of the filename), and the reason it's sending it differently is because it's normalizing one of them, and not the other.

Personally, I would just restore the behavior that twine doesn't normalize the project name, because that allows people to have control over things like is their project named "Django" or "django", and pushes the responsibility for ensuring the project name and file name matches onto the end user. However, it would work if twine just normalized the filename as well.

Reply:
I'm worried about normalizing the filename because (fortunately or not) folks use twine with servers other than PyPI and may have checks after the fact that look up the uploaded file by name. Normalizing the filename would be confusing to those users

Reply:
> I'm worried about normalizing the filename because (fortunately or not) folks use twine with servers other than PyPI and may have checks after the fact that look up the uploaded file by name. Normalizing the filename would be confusing to those users

I don't think that twine should do *any* normalization, and originally it did not, it wasn't until #736 that it started to when it started normalizing the project name.

> But that is 100% standards-conforming behaviour. The metadata contains the unnormalised name, and the filename uses a normalised name. It seems to me that it's PyPI that's wrong here, as it's not following the [spec](https://packaging.python.org/en/latest/specifications/core-metadata/#name) on how to compare names (basically, normalise before comparing).

PyPI's comparison here predates those specs :)

I'd have to think about whether there's value in the current comparison for PyPI's use case, it's currently ensuring that the filename and project name match normalization, since PyPI is blocking on ingestion here, it's fine for it to be stricter than the standards allow (and in many cases it is) [^1].

To be clear, the data that twine is sending to PyPI is the reverse of what you said, the filename is unnormalized, and the name itself is normalized.

Reply:
> Previously twine used `pkg_resources.safe_name()` to escape that field prior to sending it to PyPI, which turns `mosaik.SimConfig` into `mosaik.SimConfig` (because none of those characters are unsafe in a filename).
> 
> That commit linked changes twine to use `packaging.utils.canonicalize_name()`, which changes `mosaik.SimConfig` into `mosaik-simconfig`.
>
> I don't think that twine should do _any_ normalization, and originally it did not, it wasn't until #736 that it started to when it started normalizing the project name.

Hm. Twine is no longer using `canonicalize_name`, but rather a "vendored" copy of `safe_name`. Here's the timeline of normalization from searching the repo history:

- In https://github.com/pypa/twine/pull/70, Twine started using `pkg_resources.safe_name` to normalize the metadata name
- In https://github.com/pypa/twine/pull/736, Twine switched to using `packaging.utils.canonicalize_name` to remove the dependency on `pkg_resources`
- In https://github.com/pypa/twine/pull/745, Twine reverted to the previous normalization approach, by copying `pkg_resources.safe_name` into the `_safe_name` function

So, as near as I can tell, Twine has been normalizing the package name the same way since 2014, except for a brief period in 2021 where it used a different approach that caused issues.

Conceptually, it makes sense to me that Twine should send the metadata name unaltered to PyPI. However, that feels like a big change.

Reply:
Also, I agree that Twine shouldn't normalize filenames.

Reply:
Ahh, I see, I missed it switched back to effectively using ``safe_name()``.

So, the problem is still roughly the same, just being caused by the fact that ``safe_name`` normalizes ``_`` to ``-``.

You should not have to do *any* normalization of the name for PyPI.

Where PyPI wants things to be normalized it normalizes it itself, where it doesn't, it doesn't.

I think there was a time when that wasn't the case, and you had to do some normalization, unfortunately #70 and the linked issue #47 don't provide enough information for me to remember specifics.

If it were me, I'd be inclined to try removing all normalization from the ``name`` field and see if PyPI complains. I'm pretty sure it won't [^1], but if it does then that would be a PyPI bug IMO.

[^1]: Assuming that the unnormalized name of the project matches the filename.

Reply:
While I agree that twine should upload the file with the name as given (i.e., not normalise) I'm a bit concerned about the idea that neither twine nor the index server is validating that the files being uploaded are correct - where "correct" implies "normalised". But I guess twine can take a "trust the user" attitude, and servers might be prepared to accept unnormalised names. I'm not sure how the rest of the ecosystem would handle unnormalised wheel names (sdist names are still a bit of a mess so "legacy" forgives a lot there). I'm going to say that's not my problem though 🙂

Reply:
Difference that's splitting hairs but twine takes a trust the build-system approach followed by a trust the repository server to error with a semi-helpful error response we can provide to the user.

Genuinely, I'm not certain twine needs to exist. It provides very little benefit at this point. Most of what it does for upload can now be achieved with the standard library (historically not the case) on most (if not all of the supported versions). Most of the checks it has are simplistic and could be done by a build system to verify its output during a finalisation stage (where it would provide more value and immediate feedback to the user closer to where the issue is).

There was a dream of twine becoming a build-then-upload tool and way of hiding every build backend but neither Brian nor I have made that happen. Speaking for myself, I certainly don't have the time or desire for that. With the number of build backends, it would likely be a nightmare to paper over all their APIs (assuming they're supporting an API for us to use) and we'd have to work with whatever existing config file they're using, section, etc. Just doesn't seem worth the effort.

Reply:
There's a standard API for it, so you wouldn't need to paper over anything, just use the standard APIs... that being said, there's already a project doing that (https://github.com/pypa/build) now so adding another one seems silly.